ports/geckolib/gecko_bindings/sugar/refptr.rs

```diff
@@ -0,0 +1,91 @@
+use std::marker::PhantomData;
+use std::mem::{forget, transmute};
+use std::ptr::{read, null};
+use std::sync::Arc;
+
+/// Helper trait for conversions between FFI Strong/Borrowed types and Arcs
+pub unsafe trait RefCounted where Self: Sized {
+    /// Gecko's name for the type
+    /// This is equivalent to ArcBox<Self>
+    type FFIType: Sized;
+    fn with<F, Output>(raw: Borrowed<Self::FFIType>, cb: F) -> Output
+               where F: FnOnce(&Arc<Self>) -> Output {
+        debug_assert!(!raw.is_null());
+
+        let owned = unsafe { Self::borrowed_into(raw) };
+        let result = cb(&owned);
+        forget(owned);
+        result
+    }
+
+    fn maybe_with<F, Output>(maybe_raw: Borrowed<Self::FFIType>, cb: F) -> Output
+                         where F: FnOnce(Option<&Arc<Self>>) -> Output {
+        let owned = if maybe_raw.is_null() {
+            None
+        } else {
+            Some(unsafe { Self::borrowed_into(maybe_raw) })
+        };
+
+        let result = cb(owned.as_ref());
+        forget(owned);
+
+        result
+    }
+
+    fn into(ptr: Strong<Self::FFIType>) -> Arc<Self> {
+        unsafe { transmute(ptr) }
+    }
+
+    unsafe fn borrowed_into(ptr: Borrowed<Self::FFIType>) -> Arc<Self> {
```

Comment:
Please add doc comments explaining what all of these methods do, especially the unsafe ones, as they have contracts which aren't expressed within their type signatures and may be misunderstood.